contributors/design-proposals/troubleshoot-running-pods.md

```diff
@@ -0,0 +1,570 @@
+Debug Containers that can be created in a pod, but exceeding a pod's resource
+allocation may cause it to be evicted.
+
+### Debug Container Status
+
+The status of a *debug container* is reported in a new field in `v1.PodStatus`:
+
+```
+type PodStatus struct {
+        ...
+        DebugContainerStatuses []ContainerStatus
+}
+```
+
+This list is populated by the kubelet from
+`kubecontainer.PodStatus.ContainerStatuses` in the same way as regular and init
+container statuses. This is sent to the API server and displayed by `kubectl
+describe pod`.
+
+Status of running Debug Containers will always be reported in this list, but
+Debug Containers that exit will be removed from this list when they are garbage
+collected. In order to satisfy inspection and audit requirements, we need to
+fully understand the requirements and whether they will be satisfied by other
+audit features currently under development. Until then, Debug Containers may
+disappear from the list of Debug Container Statuses if the node is under
+resource pressure. This is acceptable for the alpha but must be resolved to
+graduate from alpha status.
+
+### Restarting and Reattaching Debug Containers
+
+Debug Containers will never be restarted automatically. It is possible to
+replace a Debug Container that has exited by re-using a Debug Container name. It
+is an error to attempt to replace a Debug Container that is still running.
+
+One way in which `kubectl debug` differs from `kubectl exec` is the ability to
+reattach to Debug Container if you've been disconnected. This is accomplished by
+running `kubectl debug -r` with the name of an existing, running Debug
+Container. The `-r` option to `kubectl debug` is translated to a `reattach=True`
+parameter. It is an error to attempt to reattach to a Debug Container that does
+not exist or is not running.
+
+Action                   | Exiting Container State | Result
+------------------------ | ----------------------- | ----------------
+Create Debug Container   | Does Not Exist          | Create & Attach
+"                        | Running                 | *ERROR*
+"                        | Exited                  | Replace & Attach
+Reattach Debug Container | Does Not Exist          | *ERROR*
+"                        | Running                 | Attach
+"                        | Exited                  | *ERROR*
+
+When supported by a runtime, multiple clients can attach to a single debug
+container and share the terminal. This is supported by Docker.
+
+### Killing Debug Containers
+
+Debug containers will not be killed automatically until the pod (specifically,
+the pod sandbox) is destroyed. Unlike `kubectl exec`, Debug Containers will not
+receive an EOF if their connection is interrupted. Instead, Debug Containers
+must be reattached to exit a running process. This could be tricky if the
+process does not allocate a TTY, in this case a second Debug Container could be
```

Comment:
@ncdc @liggitt since `kubectl debug` is mainly a tool for getting a shell with a TTY, how much do we care about output stream coordination when `kubectl` will already say "press enter to get a prompt"? I think it's reasonable to defer to a future general solution rather than trying to solve this for debug.